pbft/pbft.cpp

```diff
@@ -1041,13 +1056,443 @@ pbft::get_high_water_mark()
+                return false;
+            }
+
+            // does the sequence number, view number and hash match those of the pre prepare
+            pbft_msg prepare_msg;
+            if (!prepare_msg.ParseFromString(prepare_envelope.pbft()))
+            {
+                LOG(error) << "is_valid_viewchange_message - a prepare message is invalid";
+                return false;
+            }
+
+            if (preprepare_message.sequence() != prepare_msg.sequence() || preprepare_message.view() != prepare_msg.view() || preprepare_message.request_hash() != prepare_msg.request_hash())
+            {
+                LOG(error) << "is_valid_viewchange_message - a pre prepare message has mismatched sequence, view or request hash";
+                return false;
+            }
+
+            senders.insert(prepare_envelope.sender());
+        }
+
+        if (senders.size() < 2 * this->max_faulty_nodes() + 1)
+        {
+            LOG(error) << "is_valid_viewchange_message - not enough prepares in a prepared_proof";
+            return false;
+        }
+    }
+    return true;
+}
+
+bool
+pbft::get_sequences_and_request_hashes_from_proofs(
+        const pbft_msg& viewchange_msg
+        , std::set<std::pair<uint64_t, std::string>>& sequence_request_pairs) const
+{
+    for (int j{0}; j < viewchange_msg.prepared_proofs_size(); ++j)
+    {
+        // --- add the sequence/hash to a set
+        auto prepared_proof = viewchange_msg.prepared_proofs(j);
+
+        pbft_msg msg;
+        if (bzn_envelope envelope{prepared_proof.pre_prepare()}; !this->is_peer(envelope.sender()) || !msg.ParseFromString(envelope.pbft()))
+        {
+            return false;
+        }
+        sequence_request_pairs.insert(std::make_pair(msg.sequence(), msg.request_hash()));
+    }
+    return true;
+}
+
+bool
+pbft::is_valid_newview_message(const pbft_msg& theirs, const bzn_envelope& original_theirs) const
+{
+    // - does it contain 2f+1 viewchange messages
+    if (static_cast<size_t>(theirs.viewchange_messages_size()) < 2 * this->max_faulty_nodes() + 1)
+    {
+        return false;
+    }
+
+    std::set<std::string> viewchange_senders;
+    pbft_msg viewchange_msg;
+    for (int i{0};i < theirs.viewchange_messages_size();++i)
+    {
+        bzn_envelope original_msg{theirs.viewchange_messages(i)};
+        // - are each of those viewchange messages valid?
+        if (!viewchange_msg.ParseFromString(original_msg.pbft()) ||
+            !this->is_valid_viewchange_message(viewchange_msg, original_msg))
+        {
+            LOG(error) << "is_valid_newview_message - new view message contains invalid viewchange message";
+            return false;
+        }
+
+        if (viewchange_msg.view() != theirs.view())
+        {
+            LOG(error) << "is_valid_newview_message - a view change message has a different view than the new view message";
+            return false;
+        }
```

Comment:
check type of view change

Reply:
Added a check for the type:

if (!viewchange_msg.ParseFromString(original_msg.pbft()) || viewchange_msg.type() != PBFT_MSG_VIEWCHANGE || !this->is_valid_viewchange_message(viewchange_msg, original_msg))
        {
            LOG(error) << "is_valid_newview_message - new view message contains invalid viewchange message";
            return false;
        }